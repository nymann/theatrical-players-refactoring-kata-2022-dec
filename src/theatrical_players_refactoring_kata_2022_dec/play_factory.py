from typing import Type

from theatrical_players_refactoring_kata_2022_dec.plays.comedy import ComedyPlay
from theatrical_players_refactoring_kata_2022_dec.plays.play import Play
from theatrical_players_refactoring_kata_2022_dec.plays.tragedy import TragedyPlay


class PlayFactory:
    _play_types = {"comedy": ComedyPlay, "tragedy": TragedyPlay}

    @classmethod
    def create_play(cls, play_dict: dict, perf: dict) -> Play:
        play_type = play_dict["type"]
        try:
            play_klass: Type[Play] = cls._play_types[play_type]
        except KeyError:
            raise ValueError(f"unknown type: {play_type}")
        return play_klass(perf=perf, play_dict=play_dict)
