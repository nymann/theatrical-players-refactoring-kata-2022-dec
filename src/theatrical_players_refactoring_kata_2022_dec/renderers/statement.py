from abc import ABC
from abc import abstractmethod

from theatrical_players_refactoring_kata_2022_dec.plays.play import Play


class StatementRenderer(ABC):
    @abstractmethod
    def render(self, plays: list[Play], customer: str) -> str:
        raise NotImplementedError

    def format_as_dollars(self, amount):
        return f"${amount:0,.2f}"
