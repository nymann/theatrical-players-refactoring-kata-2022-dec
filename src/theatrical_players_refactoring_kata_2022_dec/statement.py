import math
from typing import Any


def format_as_dollars(amount):
    return f"${amount:0,.2f}"


def statement(invoice: dict[str, Any], plays: dict[str, dict[str, str]]) -> str:
    total_amount: float = 0
    volume_credits: float = 0
    result = f'Statement for {invoice["customer"]}\n'

    for perf in invoice["performances"]:
        play = plays[perf["playID"]]
        if play["type"] == "tragedy":
            this_amount = tragedy_play_price(perf=perf)
        elif play["type"] == "comedy":
            this_amount = comedy_play_price(perf=perf)
        else:
            raise ValueError(f'unknown type: {play["type"]}')

        # add volume credits
        volume_credits += max(perf["audience"] - 30, 0)
        # add extra credit for every ten comedy attendees
        if "comedy" == play["type"]:
            volume_credits += math.floor(perf["audience"] / 5)
        # print line for this order
        result += f' {play["name"]}: {format_as_dollars(this_amount/100)} ({perf["audience"]} seats)\n'
        total_amount += this_amount

    result += f"Amount owed is {format_as_dollars(total_amount/100)}\n"
    result += f"You earned {volume_credits} credits\n"
    return result


def tragedy_play_price(perf: dict) -> float:
    result = 40000
    if perf["audience"] > 30:
        result += 1000 * (perf["audience"] - 30)
    return result


def comedy_play_price(perf: dict) -> float:
    this_amount = 30000
    if perf["audience"] > 20:
        this_amount += 10000 + 500 * (perf["audience"] - 20)

    this_amount += 300 * perf["audience"]
    return this_amount
