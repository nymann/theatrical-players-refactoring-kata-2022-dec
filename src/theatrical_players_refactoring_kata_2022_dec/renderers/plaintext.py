from theatrical_players_refactoring_kata_2022_dec.plays.play import Play
from theatrical_players_refactoring_kata_2022_dec.renderers.statement import StatementRenderer
from theatrical_players_refactoring_kata_2022_dec.statement import Statement


class PlaintextStatementRenderer(StatementRenderer):
    def render(self, statement: Statement) -> str:
        result = f"Statement for {statement.customer}\n"
        for play in statement.plays:
            result += self._format_play(play=play)

        result += self._amount_owned(statement=statement)
        result += f"You earned {statement.volume_credits} credits\n"
        return result

    def _format_play(self, play: Play) -> str:
        dollar_price = self.format_as_dollars(play.price() / 100)
        return f" {play.name}: {dollar_price} ({play.audience} seats)\n"

    def _amount_owned(self, statement: Statement) -> str:
        amount_owned = self.format_as_dollars(statement.total_amount / 100)
        return f"Amount owed is {amount_owned}\n"
