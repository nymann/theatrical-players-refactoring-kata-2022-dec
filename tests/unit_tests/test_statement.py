import json

from approval_utilities.utils import get_adjacent_file
from approvaltests import verify

from theatrical_players_refactoring_kata_2022_dec.renderers.plaintext import PlaintextStatementRenderer
from theatrical_players_refactoring_kata_2022_dec.statement import Statement


def test_example_statement():
    renderer = PlaintextStatementRenderer()
    with open(get_adjacent_file("invoice.json")) as f:
        invoice = json.loads(f.read())
    with open(get_adjacent_file("plays.json")) as f:
        plays = json.loads(f.read())

    statement = Statement.from_dicts(invoice=invoice, plays=plays)
    verify(renderer.render(statement))


def test_statement_with_new_play_types():
    renderer = PlaintextStatementRenderer()
    with open(get_adjacent_file("invoice_new_plays.json")) as f:
        invoice = json.loads(f.read())
    with open(get_adjacent_file("new_plays.json")) as f:
        plays = json.loads(f.read())
    statement = Statement.from_dicts(invoice=invoice, plays=plays)
    verify(renderer.render(statement))
