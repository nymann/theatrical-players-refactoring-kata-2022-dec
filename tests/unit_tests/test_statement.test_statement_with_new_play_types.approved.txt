Statement for BigCoII
 Henry V: $724.00 (53 seats)
 As You Like It: $740.00 (55 seats)
Amount owed is $1,464.00
You earned 69 credits
